File: apps/ai-server/app/core/responses.py
"""Canonical API response builders and compatibility adapters."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from datetime import UTC, datetime
from typing import Any, NoReturn

from fastapi import status
from pydantic import BaseModel, ConfigDict, Field


def utc_now_iso() -> str:
    """Return the current UTC time in the API's ISO-8601 format."""

    return datetime.now(UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


class ErrorDetail(BaseModel):
    """One safe, client-facing validation or business error detail."""

    field: str | None = None
    code: str
    message: str

    model_config = ConfigDict(extra="forbid")


class ApiError(Exception):
    """A controlled error that can be converted into the standard envelope."""

    def __init__(
        self,
        *,
        status_code: int,
        business_code: str,
        message: str,
        trace_id: str,
        errors: Sequence[ErrorDetail] = (),
        headers: Mapping[str, str] | None = None,
    ) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.business_code = business_code
        self.message = message
        self.trace_id = trace_id
        self.errors = tuple(errors)
        self.headers = dict(headers or {})


def success_response(
    *,
    business_code: str,
    message: str,
    trace_id: str,
    data: Any = None,
    meta: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Build the canonical success response envelope."""

    return {
        "success": True,
        "businessCode": business_code,
        "message": message,
        "data": data,
        "meta": dict(meta or {}),
        "traceId": trace_id,
    }


def error_response(
    *,
    business_code: str,
    message: str,
    trace_id: str,
    errors: Sequence[ErrorDetail] = (),
    meta: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Build the canonical error envelope.

    Field-level errors are stored under ``meta.fieldErrors`` according to the
    canonical API design.  The legacy ``error_payload`` adapter below keeps
    top-level ``errors`` for existing callers during the transition.
    """

    response_meta = dict(meta or {})
    if errors:
        response_meta["fieldErrors"] = [error.model_dump(exclude_none=True) for error in errors]

    return {
        "success": False,
        "businessCode": business_code,
        "message": message,
        "data": None,
        "meta": response_meta,
        "traceId": trace_id,
    }


def raise_api_error(
    *,
    status_code: int,
    business_code: str,
    message: str,
    trace_id: str,
    errors: Sequence[ErrorDetail] = (),
    headers: Mapping[str, str] | None = None,
) -> NoReturn:
    """Raise a controlled error from a view or dependency."""

    raise ApiError(
        status_code=status_code,
        business_code=business_code,
        message=message,
        trace_id=trace_id,
        errors=errors,
        headers=headers,
    )


def success_payload(
    *,
    business_code: str,
    message: str,
    trace_id: str,
    data: Any = None,
    meta: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Compatibility alias for the original success helper."""

    return success_response(
        business_code=business_code,
        message=message,
        trace_id=trace_id,
        data=data,
        meta=meta,
    )


def error_payload(
    *,
    business_code: str,
    message: str,
    trace_id: str,
    errors: Sequence[ErrorDetail],
) -> dict[str, Any]:
    """Compatibility adapter that preserves top-level ``errors``."""

    return {
        "success": False,
        "businessCode": business_code,
        "message": message,
        "errors": [error.model_dump() for error in errors],
        "traceId": trace_id,
    }


class ApiResponse(BaseModel):
    """Compatibility model used by older module code."""

    business_code: str
    message: str
    result: Any = None
    trace_id: str
    meta: dict[str, Any] | None = None
    status_code: int = Field(
        default=status.HTTP_200_OK,
        ge=100,
        le=599,
    )

    def success_payload(self) -> dict[str, Any]:
        return success_response(
            business_code=self.business_code,
            message=self.message,
            trace_id=self.trace_id,
            data=self.result,
            meta=self.meta,
        )

    def raise_error(self) -> NoReturn:
        raise_api_error(
            status_code=self.status_code,
            business_code=self.business_code,
            message=self.message,
            trace_id=self.trace_id,
        )

File: apps/work-server/src/core/responses.ts
import type { Response } from "express";

import { getTraceId, TRACE_HEADER } from "./trace.js";
import type { Request } from "express";

export interface ErrorDetail {
  field?: string;
  code: string;
  message: string;
}

export interface ApiEnvelope<T = unknown> {
  success: boolean;
  businessCode: string;
  message: string;
  data: T | null;
  meta: Record<string, unknown>;
  traceId: string;
}

export class ApiError extends Error {
  readonly statusCode: number;
  readonly businessCode: string;
  readonly errors: ErrorDetail[];
  readonly headers: Record<string, string>;

  constructor(options: {
    statusCode: number;
    businessCode: string;
    message: string;
    errors?: ErrorDetail[];
    headers?: Record<string, string>;
  }) {
    super(options.message);
    this.name = "ApiError";
    this.statusCode = options.statusCode;
    this.businessCode = options.businessCode;
    this.errors = options.errors ?? [];
    this.headers = options.headers ?? {};
  }
}

export function jsonSafe(value: unknown): unknown {
  if (typeof value === "bigint") return Number(value);
  if (value instanceof Date) return value.toISOString();
  if (Array.isArray(value)) return value.map(jsonSafe);
  if (typeof value === "object" && value !== null) {
    return Object.fromEntries(
      Object.entries(value).map(([key, child]) => [key, jsonSafe(child)]),
    );
  }
  return value;
}

export function successResponse<T>(options: {
  request: Request;
  response: Response;
  statusCode: number;
  businessCode: string;
  message: string;
  data?: T;
  meta?: Record<string, unknown>;
}): Response {
  const traceId = getTraceId(options.request);
  options.response.setHeader(TRACE_HEADER, traceId);
  return options.response.status(options.statusCode).json({
    success: true,
    businessCode: options.businessCode,
    message: options.message,
    data: jsonSafe(options.data ?? null),
    meta: options.meta ?? {},
    traceId,
  } satisfies ApiEnvelope);
}

export function errorResponse(options: {
  request: Request;
  response: Response;
  statusCode: number;
  businessCode: string;
  message: string;
  errors?: ErrorDetail[];
  headers?: Record<string, string>;
}): Response {
  const traceId = getTraceId(options.request);
  const meta = options.errors?.length ? { fieldErrors: options.errors } : {};
  options.response.setHeader(TRACE_HEADER, traceId);
  for (const [key, value] of Object.entries(options.headers ?? {})) {
    options.response.setHeader(key, value);
  }
  return options.response.status(options.statusCode).json({
    success: false,
    businessCode: options.businessCode,
    message: options.message,
    data: null,
    meta,
    traceId,
  } satisfies ApiEnvelope<null>);
}

export function sendSuccess<T>(
  request: Request,
  response: Response,
  statusCode: number,
  businessCode: string,
  message: string,
  data?: T,
  meta?: Record<string, unknown>,
): Response {
  return successResponse({ request, response, statusCode, businessCode, message, data, meta });
}
